### DISCUSSION TITLE
Package Owner

### DISCUSSION THREAD

**Discussion by ElectronicsArchiver** [pulsar-edit/.github#107] [Pulsar Support] [answered]
<br>

TLDR

<br>

- I took over maintenance for two packages `highlight-line`, `highlight-selected`

- The owner moved the repositories to a new organization https://github.com/Pulsar-Edit-Highlights/

*( Now focusing only on `highlight-line`, though it applies for both )*

- In the future I would like to be able to publish new versions, however the owner of the package, 
  as seen on https://web.pulsar-edit.dev/packages/highlight-line is his personal account.

<br>
<br>

## Questions

- How can the owner be switched to the organization?

- What permissions does the PAT need?

**Comment by confused-Techie** [comment, marked as answer]
So this is a discussion we had previously, and with our currently planned solution, we are still open to discussion. But as it stands now here is the answer we have currently agreed on.

So ideally we wouldn't be transferring ownership of packages themselves. Because that opens a whole can of worms where someone could say they want ownership of a package, but then later if the maintainer were to come back and want it, or if someone else were to say they want ownership of the package so on and so forth. So we would hope to never manually be transferring ownership.

Now the simplest solution here would be to publish the package under a new name, and we are hoping to soon support extra flags on packages, to where if you inform us that your new package under whatever name is the updated and maintained fork of an already existing package, then we could inform the user that the old package is no longer updated, and to use yours instead. The same way that NPM will commonly do when installing older packages.

Now the interesting thing here is it's still essentially the same package, so I would be interested honestly if the backend would figure it out when publishing a new version. There is baked in support for trying to work around a package rename, but I can't say with any certainty that it would successfully handle a request when being published by a new author, on a different repo.

So we could discuss further but I personally feel the right answer is to publish a brand new package under a new name, and once supported we could add a flag to it.

---

As for your question on the permissions of the PAT token, all that's recommended to enable is the following:

* public_repo
* read:org

You could assign more, but that's all that the backend will use. This is something I thought was in the documentation linked from the signup page, but I'll double check and see what I can do to add it there.

**Reply by ElectronicsArchiver** [reply]
Thank you for the answer.

In regards to linking to a new fork, I don't see it as a viable solution
considering this is the original project, the owner just didn't want to
be burdened by the work anymore and I imagine this situation might 
come up more times in the future.

Just linking to a new fork doesn't cut it since:

- If there isn't an easy way like a `switch to` button, few users would actually consider switching ( and in this case, just updating )

- Without a way to disable the old package from search, both old and new will show up and be confusing to the user, especially in this case where it's **the same package, not even a fork**

- I'd would be ideal to have a `Transfer` button, just like GitHub does for repositories, that allows the current owner to assign a new one, such as in this case, where he would assign it from his personal account to his organization. And in regards to who can transfer when it belongs to an org, just allow users that are owners of the org to do it.

**Reply by confused-Techie** [reply]
@ElectronicsArchiver 
So we have found an a way to delist packages if needed, so we could viably delist the older version of a package and just leave the most up to date version there instead.

I do get what you mean that it may not be the best solution. But you are right if we do support it, it has to be done by the original owner. Since otherwise it's just a managerial and security headache.

To be perfectly honest our integration when it comes to GitHub and especially confirming ownership of packages and repositories is still something that's being worked on nearly everyday.

So I'll add this item onto the package backend feature request [Project Board](https://github.com/orgs/pulsar-edit/projects/2) and I'd recommend you keep an eye on it there, but this may be a feature that takes some time to implement.

The hardest thing about this, is because the backend doesn't store any credentials of a user it can sometimes be difficult to confirm ownership of a package when it's data has changed, such as the GitHub repo name and/or ORG. 

Since for example in this case, if the package was originally published from an org, anyone within the org (or with read and write access really) would be able to publish package versions for the package. But in this case, it's now the question of verifying that someone else can take ownership of a package who's Author has essentially changed, which in that case you can see how it becomes much more difficult to confirm everything.

But like I said, I'll add it to our feature requests and see what we can do.

Although in the short term if you want this package published right now it may be best to publish under a new name, and if wanted could even ask the original author to delete their package.

**Reply by ElectronicsArchiver** [reply]
Thanks for adding it to the project board.

Btw, do you have a way to publish ( small )
updates yet, considering the `blog` area on
the website has 3 posts?

If not, why not just append a list at the end
of the organizations overview page, there all
devs, no matter what socials they use could be
informed about small changes like a feature
being ready or being worked on or the like ..

**Reply by confused-Techie** [reply]
Okay wait, I originally started typing a whole bunch before I got what you mean.

Are you referring to similar how we have on Discord the `#announcements` channel where we put small little updates about whats new around the ecosystem. You know this is a really good idea. It does seem that Discord or Mastodoon or even Reddit I suppose would be the right form factor for it.

Hmm, but you are right it should be agnostic in terms of what platform others use to get ahold of those updates. I almost wonder if a Discord bot could take everything we put in that Discord channel and generate an RSS feed. Then we could plug it into the website, or users could just add it to whatever RSS Feed Reader they prefer, and that seems like the most agnostic way to do it while still being a proper platform.

Since GitHub feels much more long form to me at least for those small 1 sentence updates. The sort of thing I'd personally rather see buried in a ChangeLog somewhere other than be displayed prominently 

But I actually do like this idea, when I'm off work I'll try and see what the team thinks

**Reply by ElectronicsArchiver** [reply]
I don't have any experience with RSS feeds, but it sounds good.

What I was thinking of was a simple yaml file with 'news' in it
with a workflow that automatically styles and inserts them into
the organizations overview ( `.GitHub/profiles/README.md` )

I'm actually working on that ( because I felt like it ),
I will ping you once it's done ( even if you don't plan
on using that method )
